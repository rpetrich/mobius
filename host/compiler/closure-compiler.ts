export default function closure() {
	return {
		name: "closure-compiler",
		transformBundle(src: string) {
			const { compile } = require("google-closure-compiler-js");
			const output = compile({
				languageIn: "ES5",
				languageOut: "ES5",
				assumeFunctionWrapper: false,
				rewritePolyfills: false,
				createSourceMap: true,
				processCommonJsModules: true,
				jsCode: [{ src }],
			});
			return {
				code: output.compiledCode,
				map: output.sourceMap,
			};
		},
	};
}
