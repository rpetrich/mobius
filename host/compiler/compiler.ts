import { readFileSync } from "fs";
import { resolve } from "path";
import { cwd } from "process";
import * as ts from "typescript";
import { exists, mkdir, modifiedTime, packageRelative, readJSON, writeFile } from "../fileUtils";
import { typescript } from "../lazy-modules";
import { virtualModule } from "../lazy-modules";
import { once } from "../memoize";
import { VirtualModule } from "../modules/index";

function loadCompilerOptions(mode: "server" | "client", mainPath: string) {
	const basePath = resolve(mainPath, "..");
	const fileName = `tsconfig-${mode}.json`;
	const configFile = typescript.readJsonConfigFile(packageRelative(fileName), (path: string) => readFileSync(path).toString());
	const configObject = typescript.convertToObject(configFile, []);
	const result = typescript.convertCompilerOptionsFromJson(configObject.compilerOptions, packageRelative("./"), fileName).options;
	result.suppressOutputPathCheck = true;
	result.baseUrl = basePath;
	result.paths = {
		"app": [
			mainPath,
		],
		"*": [
			packageRelative(`${mode}/*`),
			resolve(basePath, `${mode}/*`),
			packageRelative("common/*"),
			resolve(basePath, "common/*"),
			packageRelative("types/*"),
			resolve(basePath, "*"),
		],
		"tslib": [
			packageRelative("node_modules/tslib/tslib"),
		],
		"babel-plugin-transform-async-to-promises/helpers": [
			resolve(require.resolve("babel-plugin-transform-async-to-promises"), "../helpers"),
		],
		"preact": [
			packageRelative("dist/common/preact"),
		],
		"redom": [
			packageRelative("host/redom"),
		],
	};
	return result;
}

const diagnosticsHost = {
	getCurrentDirectory: cwd,
	getCanonicalFileName(fileName: string) {
		return fileName;
	},
	getNewLine() {
		return "\n";
	},
};

const declarationPattern = /\.d\.ts$/;

const getDocumentRegistry = once(() => typescript.createDocumentRegistry());

export interface CompiledOutput<T> {
	program: ts.Program;
	compiler: Compiler<T>;
	resolveModule(moduleName: string, containingFile: string): { resolvedFileName: string, isExternalLibraryImport?: boolean } | void;
	getAllFileNames(): string[];
	getEmitOutput(path: string): { code: string, map: string | undefined } | void;
	getVirtualModule(path: string): VirtualModule | void;
	saveCache(newCache: T): Promise<void>;
}

const compilerModifiedTime = once(() => modifiedTime(packageRelative("dist/host/compiler/compiler.js")));

interface VirtualModuleCacheEntry {
	dependencies: { [dependency: string]: number };
	declaration?: string;
	code?: string;
}

export interface CacheData<T> {
	path: string;
	data?: T;
	virtualModules: { [module: string]: VirtualModuleCacheEntry };
}

const version: number = 1;

export async function loadCache<T>(mainPath: string, cacheSlot: string): Promise<CacheData<T>> {
	const path = resolve(mainPath, `../.cache/${cacheSlot}.json`);
	if (await exists(path) && (modifiedTime(path) > compilerModifiedTime())) {
		const result = await readJSON(path);
		if (result && result.version === version) {
			return {
				path,
				data: result.data,
				virtualModules: result.virtualModules || { },
			};
		}
	}
	return {
		path,
		virtualModules: { },
	};
}

export function noCache<T>(): CacheData<T> {
	return {
		path: "",
		virtualModules: { },
	};
}

function notExternal<T extends ts.ResolvedModule | undefined>(resolvedModule: T): T {
	// Override to allow paths inside node_modules/mobius-js/ to be emitted by TypeScript
	if (resolvedModule && !/\.d\.ts$/.test(resolvedModule.resolvedFileName)) {
		resolvedModule.isExternalLibraryImport = false;
	}
	return resolvedModule;
}

export class Compiler<T> {
	public readonly basePath: string;
	public readonly compilerOptions: ts.CompilerOptions;
	private readonly paths: string[];
	private readonly host: ts.LanguageServiceHost & ts.ModuleResolutionHost;
	private readonly languageService: ts.LanguageService;
	private readonly resolutionCache: ts.ModuleResolutionCache;
	private virtualModules: { [path: string]: VirtualModule } = { };
	private virtualModuleEntries: { [module: string]: VirtualModuleCacheEntry } = { };

	constructor(mode: "server" | "client", public readonly cache: CacheData<T>, mainPath: string, rootFileNames: string[], private minify: boolean, private fileRead: (path: string) => void) {
		this.basePath = resolve(mainPath, "..");
		this.compilerOptions = loadCompilerOptions(mode, mainPath);
		this.paths = (require("module")._nodeModulePaths(this.basePath) as string[]).concat(module.paths);
		const readFile = (path: string, encoding?: string) => {
			const module = this.getVirtualModule(path);
			if (module) {
				return module.generateTypeDeclaration();
			}
			fileRead(path);
			return typescript.sys.readFile(path, encoding);
		};
		this.resolutionCache = typescript.createModuleResolutionCache(this.basePath, (s) => s);
		this.host = {
			getScriptFileNames() {
				return rootFileNames;
			},
			getScriptVersion: (fileName) => {
				return modifiedTime(fileName).toString();
			},
			getScriptSnapshot(fileName) {
				const contents = readFile(fileName);
				if (typeof contents !== "undefined") {
					return typescript.ScriptSnapshot.fromString(contents);
				}
				return undefined;
			},
			getCurrentDirectory() {
				return typescript.sys.getCurrentDirectory();
			},
			getCompilationSettings: () => {
				return this.compilerOptions;
			},
			getDefaultLibFileName(options) {
				return typescript.getDefaultLibFilePath(options);
			},
			readFile,
			fileExists: (path: string) => {
				const result = typescript.sys.fileExists(path);
				if (result) {
					return result;
				}
				if (this.getVirtualModule(path)) {
					return true;
				}
				return false;
			},
			readDirectory: typescript.sys.readDirectory,
			directoryExists(directoryName: string): boolean {
				return typescript.sys.directoryExists(directoryName);
			},
			getDirectories(directoryName: string): string[] {
				return typescript.sys.getDirectories(directoryName);
			},
			resolveModuleNames: (moduleNames: string[], containingFile: string, reusedNames?: string[]) => {
				return moduleNames.map((moduleName) => notExternal(typescript.resolveModuleName(moduleName, containingFile, this.compilerOptions, this.host, this.resolutionCache).resolvedModule!));
			},
			getResolvedModuleWithFailedLookupLocationsFromCache: (moduleName: string, containingFile: string) => {
				const result = typescript.resolveModuleName(moduleName, containingFile, this.compilerOptions, this.host, this.resolutionCache);
				notExternal(result.resolvedModule);
				return result;
			},
		};
		this.languageService = typescript.createLanguageService(this.host, getDocumentRegistry());
	}

	public fileChanged = (path: string) => {
		/* tslint:disable no-empty */
	}

	private getVirtualModule = (path: string) => {
		if (declarationPattern.test(path)) {
			path = path.replace(declarationPattern, "");
			if (Object.hasOwnProperty.call(this.virtualModules, path)) {
				return this.virtualModules[path];
			}
			const entry: VirtualModuleCacheEntry = {
				dependencies: { },
			};
			const result = virtualModule.default(this.basePath, path, this.minify, (filePath: string) => {
				entry.dependencies[filePath] = modifiedTime(filePath);
				this.fileRead(path);
			}, this.compilerOptions);
			if (result) {
				this.virtualModuleEntries[path] = entry;
				if (this.cache && this.cache.virtualModules && Object.hasOwnProperty.call(this.cache.virtualModules, path)) {
					const pendingCacheEntry = this.cache.virtualModules[path];
					let stale = false;
					for (const dependency in pendingCacheEntry.dependencies) {
						if (Object.hasOwnProperty.call(pendingCacheEntry.dependencies, dependency)) {
							if (modifiedTime(dependency) !== pendingCacheEntry.dependencies[dependency]) {
								stale = true;
								break;
							}
						}
					}
					if (!stale) {
						entry.dependencies = Object.assign({}, pendingCacheEntry.dependencies);
						if (typeof pendingCacheEntry.declaration !== "undefined") {
							entry.declaration = pendingCacheEntry.declaration;
						}
						if (typeof pendingCacheEntry.code !== "undefined") {
							entry.code = pendingCacheEntry.code;
						}
					}
				}
				// Declaration is eager
				const declaration = entry.declaration || (entry.declaration = result.generateTypeDeclaration());
				return this.virtualModules[path] = {
					generateTypeDeclaration() {
						return declaration;
					},
					generateModule() {
						// Module is lazy
						return entry.code || (entry.code = result.generateModule());
					},
					instantiateModule: result.instantiateModule,
					generateStyles: result.generateStyles,
				};
			}
		}
	}

	public compile(): CompiledOutput<T> {
		this.virtualModules = { };
		const program = this.languageService.getProgram();
		if (typeof program === "undefined") {
			throw new Error("Expected language service to have a program!");
		}
		const diagnostics = typescript.getPreEmitDiagnostics(program);
		if (diagnostics.length) {
			console.log(typescript.formatDiagnostics(diagnostics, diagnosticsHost));
		}
		return {
			program,
			compiler: this,
			resolveModule: (moduleName: string, containingFile: string): { resolvedFileName: string, isExternalLibraryImport?: boolean } | void => {
				const tsResult = typescript.resolveModuleName(moduleName, containingFile, this.compilerOptions, this.host, this.resolutionCache).resolvedModule;
				if (tsResult && !tsResult.isExternalLibraryImport) {
					if (tsResult.extension === ".d.ts") {
						const replaced = tsResult.resolvedFileName.replace(declarationPattern, ".js");
						if (this.host.fileExists(replaced)) {
							return {
								resolvedFileName: replaced,
								isExternalLibraryImport: tsResult.isExternalLibraryImport,
							};
						}
					}
					return {
						resolvedFileName: tsResult.resolvedFileName,
						isExternalLibraryImport: tsResult.isExternalLibraryImport,
					};
				}
				return {
					resolvedFileName: (module.constructor as any)._resolveFilename(moduleName, null, false, { paths: this.paths }) as string,
					isExternalLibraryImport: true,
				};
			},
			getAllFileNames() {
				return program.getSourceFiles().map(value => value.fileName);
			},
			getEmitOutput: (path: string): { code: string, map: string | undefined } | void => {
				const sourceFile = program.getSourceFile(path);
				if (sourceFile) {
					let code: string | undefined;
					let map: string | undefined;
					for (const { name, text } of this.languageService.getEmitOutput(path).outputFiles) {
						if (/\.js$/.test(name)) {
							code = text;
						} else if (/\.js\.map$/.test(name)) {
							map = text;
						}
					}
					if (typeof code === "string") {
						return { code, map };
					}
				}
			},
			getVirtualModule: this.getVirtualModule,
			saveCache: async (data: T) => {
				this.cache.data = data;
				this.cache.virtualModules = this.virtualModuleEntries;
				this.virtualModuleEntries = { };
				const path = this.cache.path;
				if (path) {
					const parentPath = resolve(path, "..");
					if (!await exists(parentPath)) {
						await mkdir(parentPath);
					}
					await writeFile(path, JSON.stringify({ version, data, virtualModules: this.cache.virtualModules }));
				}
			},
		};
	}

}
