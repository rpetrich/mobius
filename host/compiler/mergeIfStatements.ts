import { NodePath } from "babel-traverse";
import { IfStatement, isBooleanLiteral, isIfStatement, isReturnStatement, logicalExpression, Node } from "babel-types";

function isReturnFalseIfStatement(node: Node): node is IfStatement {
	return isIfStatement(node) && !node.alternate && isReturnStatement(node.consequent) && isBooleanLiteral(node.consequent.argument) && !node.consequent.argument.value;
}

export default function() {
	return {
		visitor: {
			IfStatement(path: NodePath<IfStatement>) {
				const container = path.container;
				if (typeof (container as any).length == "number" && isReturnFalseIfStatement(path.node)) {
					const index = (container as any[]).indexOf(path.node);
					if (index > 0) {
						const previous = path.getSibling((index - 1) as any as string);
						if (previous.isIfStatement() && isReturnFalseIfStatement(previous.node)) {
							previous.get("test").replaceWith(logicalExpression("||", previous.node.test, path.node.test));
							path.remove();
						}
					}
				}
			},
		},
	};
}
