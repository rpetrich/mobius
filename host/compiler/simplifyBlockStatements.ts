import { NodePath } from "babel-traverse";
import { BlockStatement } from "babel-types";

export default function() {
	return {
		visitor: {
			BlockStatement(path: NodePath<BlockStatement>) {
				if (typeof (path.container as any).length == "number" && path.node.body.length) {
					path.replaceWithMultiple(path.node.body);
				}
			},
		},
	};
}
