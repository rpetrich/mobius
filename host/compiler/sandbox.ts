import { readFileSync } from "fs";
import { RawSourceMap } from "source-map";
import * as vm from "vm";
import { modifiedTime } from "../fileUtils";
import memoize from "../memoize";
import { ModuleMap, StaticAssets } from "../modules/index";
import { LocalSessionSandbox } from "../session-sandbox";
import { CompiledOutput } from "./compiler";

export interface ModuleSource {
	path: string;
	sandbox: boolean;
}

export interface ServerModule {
	exports: any;
	paths: string[];
}

export interface ServerModuleGlobal {
	self: this;
	global: this | NodeJS.Global;
	require: (name: string) => any;
	module: ServerModule;
	exports: any;
	Object?: typeof Object;
	Array?: typeof Array;
}

export interface LoaderCacheData {
	modules: { [path: string]: { initializer: string, shared: boolean, modified: number } };
}

const requireOnce = memoize(require);

function wrapSource(code: string) {
	return `(function(self){return(function(self,global,require,document,exports,Math,Date,setInterval,clearInterval,setTimeout,clearTimeout){${code}\n})(self,self.global,self.require,self.document,self.exports,self.Math,self.Date,self.setInterval,self.clearInterval,self.setTimeout,self.clearTimeout)})`;
}

function initializerStringForOutput(code: string, map: string | undefined, filename: string, shared: boolean, coverage: boolean): string {
	const babel = requireOnce("babel-core");
	// Apply babel transformation passes
	const convertToCommonJS = requireOnce("babel-plugin-transform-es2015-modules-commonjs");
	const optimizeClosuresInRender = requireOnce("babel-plugin-optimize-closures-in-render");
	const dynamicImport = requireOnce("babel-plugin-syntax-dynamic-import");
	const transformAsyncToPromises = requireOnce("babel-plugin-transform-async-to-promises");
	const noImpureGetters = requireOnce("./noImpureGetters").default;
	const rewriteDynamicImport = requireOnce("./rewriteDynamicImport").default;
	let inputSourceMap: RawSourceMap | undefined;
	if (typeof map === "string") {
		// Patch up input source map to have proper paths
		inputSourceMap = JSON.parse(map);
		inputSourceMap!.file = filename;
		inputSourceMap!.sources[0] = filename;
	}
	const additionalPlugins = coverage ? [[requireOnce("babel-plugin-istanbul").default, { filename }]] : [];
	if (shared) {
		const singlePass = babel.transform(code, {
			babelrc: false,
			compact: false,
			plugins: additionalPlugins.concat([
				dynamicImport,
				rewriteDynamicImport,
				[convertToCommonJS, { noInterop: true }],
				noImpureGetters,
				[transformAsyncToPromises, { externalHelpers: true, hoist: true }],
				optimizeClosuresInRender,
			]),
			inputSourceMap,
			filename,
		});
		return `(function(require){return ${wrapSource(singlePass.code!)}\n})`;
	} else {
		const firstPass = babel.transform(code, {
			babelrc: false,
			compact: false,
			plugins: additionalPlugins.concat([
				dynamicImport,
				rewriteDynamicImport,
				[convertToCommonJS, { noInterop: true }],
				noImpureGetters,
			]),
			inputSourceMap,
			filename,
		});
		const hoistSharedLabels = requireOnce("./hoistSharedLabels").default;
		const secondPass = babel.transform("return " + wrapSource(firstPass.code!), {
			babelrc: false,
			compact: false,
			plugins: [
				[convertToCommonJS, { noInterop: true }],
				[transformAsyncToPromises, { externalHelpers: true, hoist: true }],
				optimizeClosuresInRender,
				hoistSharedLabels,
			],
			parserOpts: {
				allowReturnOutsideFunction: true,
			},
			filename,
		});
		return `(function(require){${secondPass.code!}\n})`;
	}
}

export type ModuleLoader = (source: ModuleSource, module: ServerModule, globalProperties: any, sandbox: LocalSessionSandbox, require: (name: string) => any) => void;

export function sandboxLoaderForOutput(compiled: CompiledOutput<LoaderCacheData>, moduleMap: ModuleMap, staticAssets: StaticAssets, coverage: boolean): ModuleLoader {
	const loadersForPath = new Map<string, (module: ServerModule, globalProperties: any, sandbox: LocalSessionSandbox, require: (name: string) => any) => void>();

	// Extract compiled output and source map from TypeScript
	const modules: { [path: string]: { initializer: string, shared: boolean, modified: number } } = { };
	const cache = compiled.compiler.cache.data;
	function importOutputAtPath(path: string) {
		const modified = modifiedTime(path);
		if (!isNaN(modified)) {
			if (cache &&
				cache.modules &&
				Object.hasOwnProperty.call(cache.modules, path) &&
				cache.modules[path].modified === modified
			) {
				return modules[path] = cache.modules[path];
			}
			const { code, map } = compiled.getEmitOutput(path) || { code: readFileSync(path).toString(), map: undefined };
			const shared = /\/\*\*\s*@mobius:shared\s*\*\//.test(code) || /\bbabel-plugin-transform-async-to-promises\/helpers\b/.test(path) || /\bdist\/common\/preact\b/.test(path);
			return modules[path] = {
				initializer: initializerStringForOutput(code, map, path, shared, coverage),
				shared,
				modified,
			};
		}
	}
	for (const { fileName, isDeclarationFile } of compiled.program.getSourceFiles()) {
		importOutputAtPath(isDeclarationFile ? fileName.replace(/\.d\.ts$/, ".js") : fileName);
	}

	// Should find a way of doing this cleanly that properly waits for all async module loads in initial render
	setTimeout(() => compiled.saveCache({ modules }), 200);

	function initializerForPath(path: string, staticRequire: (name: string) => any): [((global: ServerModuleGlobal, sandbox: LocalSessionSandbox) => void) | undefined, boolean] {
		// Check for virtual modules
		const module = compiled.getVirtualModule(path);
		if (module) {
			const instantiate = module.instantiateModule(moduleMap, staticAssets);
			return [instantiate, false];
		}
		// Incrementally handle files that TypeScript didn't compile, as they're discovered
		let output = modules[path];
		if (!output) {
			output = importOutputAtPath(path)!;
		}
		// Wrap in the sandbox JavaScript
		return [vm.runInThisContext(output.initializer, {
			filename: path,
			lineOffset: 0,
			displayErrors: true,
		})(staticRequire) as (global: any) => void, output.shared];
	}

	return (source: ModuleSource, module: ServerModule, globalProperties: any, sandbox: LocalSessionSandbox, require: (name: string) => any) => {
		// Create a sandbox with exports for the provided module
		const path = source.path;
		let result = loadersForPath.get(path);
		if (!result) {
			const [initializer, isShared] = initializerForPath(path, require);
			if (initializer) {
				const constructModule = function(currentModule: ServerModule, currentGlobalProperties: any, sandbox: LocalSessionSandbox, currentRequire: (name: string) => any) {
					const moduleGlobal: ServerModuleGlobal & any = Object.create(global);
					Object.assign(moduleGlobal, currentGlobalProperties);
					moduleGlobal.self = moduleGlobal;
					moduleGlobal.global = global;
					moduleGlobal.require = currentRequire;
					moduleGlobal.module = currentModule;
					moduleGlobal.exports = currentModule.exports;
					initializer(moduleGlobal, sandbox);
					return moduleGlobal;
				};
				if (source.sandbox && !isShared) {
					result = constructModule;
				} else {
					const staticModule = constructModule(module, globalProperties, sandbox, require);
					result = () => staticModule;
				}
			} else {
				result = () => {
					throw new Error("Unable to find module: " + path);
				};
			}
			loadersForPath.set(path, result);
		}
		return result(module, globalProperties, sandbox, require);
	};
}
