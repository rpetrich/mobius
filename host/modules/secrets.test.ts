import { VirtualModule } from "./index";
import secrets from "./secrets";
import { ServerModuleGlobal } from "../compiler/sandbox";
import { packageRelative } from "../fileUtils";

describe("integration test", () => {

	describe("secrets/secrets.json", () => {
		const virtualModule = secrets(packageRelative("tests/secrets"), packageRelative("tests/secrets/secrets"), false, () => {}) as VirtualModule;

		describe("generated type declaration", () => {
			it("should match expected source", () => {
				expect(virtualModule.generateTypeDeclaration()).toMatchSnapshot();
			});
		});

		describe("generated module", () => {
			it("should match expected source", () => {
				expect(virtualModule.generateModule()).toMatchSnapshot();
			});
		});

		describe("instantiated module", () => {
			const fakeGlobal: ServerModuleGlobal = { self: null as any as ServerModuleGlobal, global, require, module, exports: {} };
			fakeGlobal.self = fakeGlobal;
			// Second argument isn't used by secrets pseudo-module
			virtualModule.instantiateModule({}, {})(fakeGlobal, null as any);
			it("should have a foo property", () => {
				expect(fakeGlobal.exports.foo).toBe("bar");
			});
			it("should have an integer property", () => {
				expect(fakeGlobal.exports.integer).toBe(42);
			});
		});

	});

});
