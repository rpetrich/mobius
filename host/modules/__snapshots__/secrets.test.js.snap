// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration test secrets/secrets.json generated module should match expected source 1`] = `
"import { redact } from \\"redact\\";

export const foo = redact();
export const integer = redact();"
`;

exports[`integration test secrets/secrets.json generated type declaration should match expected source 1`] = `
"import { Redacted } from \\"redact\\";

export const foo: Redacted<\\"bar\\">;
export const integer: Redacted<42>;"
`;
